### ISSUE TITLE
Store Webauthn Attestation on registration

### ISSUE THREAD

**Issue by coelho** [zitadel/zitadel#4758]
**Description**

Zitadel should store the Webauthn Attestation for future audits.
Follow up on: https://github.com/zitadel/zitadel/discussions/4429

**Acceptance criteria**
- [ ] Attestation Conveyance Preference should be configurable or be by default "direct". Currently it is "none".
https://github.com/zitadel/zitadel/blob/main/internal/webauthn/webauthn.go#L78
This appears to cause Yubikey hardware tokens to produce an empty `aaguid`.

- [ ] Raw Credential Data from registration should be stored.
https://github.com/zitadel/zitadel/blob/main/internal/webauthn/webauthn.go#L100
`credentialData` can serialize into JSON and store as-is. This is for future proofing. Currently, the `credential` var contains a very small submit of the data in `credentialData`.

**Future Work**
- Allow this information to be viewed in the console.
- Allow the administrator to whitelist tokens by `aaguid`. The database for all hardware token aaguids is also publicly available, so there could also be a directory?

**Comment by hifabienne** [comment]
Hi @coelho 
Thanks for creating the issue, we will have a look at it in our next planning and prioritize it.

**Comment by udf2457** [comment]
@hifabienne I was doing some research on ZITADEL and using the Yubikey attestation is on my checklist.  Has there been any movement on this feature since 2022 ?

**Comment by hifabienne** [comment]
Hei @udf2457 
At the moment there is not progress on this issue.
If there is a significant demand from customers/community, we will carefully consider implementing the feature.
Currently, the issue is added to our product backlog to collect feedback.

**Comment by coelho** [comment]
@hifabienne 
> Hei @udf2457 At the moment there is not progress on this issue. If there is a significant demand from customers/community, we will carefully consider implementing the feature. Currently, the issue is added to our product backlog to collect feedback.

FYI this feature is 2-fold:
1. Collect the data (tiny scope) (<- this issue)
2. Use the data (larger scope, but users can already utilize via SQL audits)
When 1. is done, a customer would have to remove every security key, and have users re-enroll, to collect the data. 

We decided to not consider a migration to Zitadel without this feature due to:
a. security posture requiring us to use and audit security keys.
b. migrating our users, enrolling security keys manually, and then having to do it all over again (when/if this is implemented) to collect and input the data into the ledger.

**Comment by hifabienne** [comment]
@livio-a @muhlemmer what is your take on this?

**Comment by udf2457** [comment]
@coelho Thank you for your detailed reply.

However I should perhaps point out that whilst you may see Passkeys as a suitable alternative, they really are not for security conscious environments.

This is because there is no attestation available with Passkeys, this is (sadly !) by design.  This means that:

1. You have no way of knowing if a Passkey was generated insecurely in software emulation or securely in an enclave
2. You have no attestation validation options to validate the devices the Passkeys are being used on

This is why Yubikey attestation has a place in the world, because:

1. Through attestation, you can validate it was securely generated on a Yubikey in non-exfilterable form
2. If you wish, you can go one level further and verify it was on a specific Yubikey hardware family, or indeed a user-specific Yubikey.

None of that can be done on Passkeys.

**Comment by coelho** [comment]
@hifabienne By "Security key" in my reply, I mean "YubiKey"/"HSM".
- I agree, Passkeys are an unfortunate mistake. 
- I agree, Webauthn attestation is the only way to ensure the use of hardware keys / secure enclaves. Otherwise, there is no way to audit or guarantee that 2FA/Passwordless is setup securely.

**Comment by elinashoko** [comment]
Thanks for filing this issue. As part of an ongoing backlog cleanup, we're closing issues that have had no activity in over 12 months.

If this is still relevant, let us know in a comment (or reopen the issue) and we'll take another look.

This issue will close in 30 days if there's no further activity.

**Comment by udf2457** [comment]
I have to say, I really dislike maintainers who use GH stale-bot to sweep things under the carpet.

Just because issues have had no activity for a long time it does not mean the issue has magically disappeared.

In this instance @elinashoko is acting as human stale-bot "as part of an ongoing backlog cleanup" is basically "we are sweeping this under the carpet".

it is even worse in the case of this issue because just look .... @muhlemmer  added it to "engineering kanban" only last month !  June 24.

**Comment by elinashoko** [comment]
I agree that inactivity doesn’t mean an issue has magically disappeared. The `to-be-closed` label doesn’t claim that it has - it identifies the issue for review as part of cleaning up a backlog of thousands of issues.

Nothing is being automatically or immediately closed. Every labelled issue receives a comment and remains open for 30 days. If someone comments on or upvotes it during that period, the label is automatically removed and the product team reviews it.

The engineering kanban addition on June 24 wasn’t a decision to prioritise or actively work on this particular issue. We recently switched our engineering workflow from sprints to kanban, which added all existing issues to the new board. Project-board changes therefore aren’t counted as issue activity.

I’ve shared the full cleanup process and rationale here: https://discord.com/channels/927474939156643850/927865053242335282/1531319896909287564

You’re welcome to comment on the issue if it still affects you or you believe it should remain open - that’s precisely why the 30-day review period exists. But describing a careful and transparent review process as “sweeping things under the carpet” isn’t a fair representation of what we’re doing.

**Comment by udf2457** [comment]
>Nothing is being automatically or immediately closed. Every labelled issue receives a comment and remains open for 30 days. If someone comments on or upvotes it during that period, the label is automatically removed and the product team reviews it.

But this is precisely the same workflow as github stale-bot carpet-sweeping ?  Bot labels it, someone comments to "save" it , label removed, theoretically the team reviews it, but they don't, the bot comes back, mark as stale again etc. etc. ..... rinse and repeat...

Of course its your project and you can do what you like, but closing things because you don’t have time to get to them leaves a bad taste in people’s mouths.  That is what you are doing here and that is what GH stale-bot does too.

**Comment by elinashoko** [comment]
I understand the comparison, but I think we disagree on what an open backlog is meant to represent.

A valid request doesn’t automatically belong in the product backlog indefinitely. Not every feature request fits Zitadel’s direction, and not every improvement to an older or soon-to-be-replaced part of the product is something we should invest in. Prioritisation means deliberately deciding what belongs in an identity platform that needs to remain coherent, secure and maintainable for millions of users.

Closing an issue doesn’t erase the problem or the discussion. It means we’re being honest that it isn’t something we currently plan to pursue. In my view, leaving thousands of issues open for years when there is no realistic intention of addressing them would be less transparent.

If renewed activity shows that an issue still matters, the product team will review it. That doesn’t guarantee it stays open forever, but it ensures we consider the current context before deciding what to do with it.

**Comment by coelho** [comment]
@elinashoko 
1. You’re unlikely to receive much feedback on this issue because organizations impacted by it have already ruled out Zitadel. For those customers, this limitation prevents Zitadel from meeting their compliance requirements and prevents implementation whatsoever.
2. If the Zitadel team doesn’t see the value in implementing this feature, moving on was a reasonable decision. Security doesn't seem to be a priority here.

**Comment by elinashoko** [comment]
Thank you for the feedback. I understand that some organisations may have ruled out Zitadel without commenting here, so GitHub engagement can’t give us the complete picture. At the same time, we can’t prioritise features based on demand we have no evidence of.

This is a valid but fairly specific compliance requirement, and we have to weigh it against other security work and customer needs. If you know of organisations currently blocked by it, concrete examples would genuinely be useful.

I don’t think it’s fair to conclude that security isn’t a priority based on this particular feature not being prioritised. Security is much broader than any single feature request, and every security or compliance-related capability still has to be considered alongside other needs.
